frame/trading/src/lib.rs

```diff
@@ -0,0 +1,239 @@
+pub type MomentOf<T> = <<T as Trait>::Time as Time>::Moment;
+pub type OfferOf<T> = Offer<<T as system::Trait>::AccountId, BalanceOf<T>, MomentOf<T>>;
+
+const TRADING_ID: LockIdentifier = *b"trading_";
+
+/// The module's configuration trait.
+pub trait Trait: system::Trait {
+    // use amount of values.
+    type Currency: LockableCurrency<Self::AccountId, Moment = Self::BlockNumber>;
+    /// Time used calculating expired.
+    type Time: Time;
+    /// The helper of checking the state of operators.
+    type OperatorFinder: OperatorFinder<Self::AccountId>;
+    /// The helper of transfering operator's authorities.
+    type TransferOperator: TransferOperator<Self::AccountId>;
+    /// The overarching event type.
+    type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait> as Operator {
+        /// A mapping from the offering account id to Offer
+        pub Offers: linked_map T::AccountId => Option<OfferOf<T>>;
+    }
+}
+
+decl_module! {
+    /// The module declaration.
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        // Initializing events
+        // this is needed only if you are using events in your module
+        fn deposit_event() = default;
+
+        /// Offer is an easy cotnract to trade.
+        /// If the sender `accept` during the period, the operator trading will be completed.
+        /// After the offer, the part of the amount of the buyer's balances will lock.
+        ///
+        /// Note: Only one offer can be issued at the same time each an account.
+        pub fn offer(origin, sender: T::AccountId, contracts: Vec<T::AccountId>, amount: BalanceOf<T>, expired: MomentOf<T>) {
+            let buyer = ensure_signed(origin)?;
+            let offer_account = buyer.clone();
+
+            if T::Currency::free_balance(&buyer) <= amount {
+                Err("buyer does not have enough balances.")?
+            }
+            // check that the operator has contracts.
+            let has_contracts = T::OperatorFinder::contracts(&sender);
+            if !contracts.iter().all(|v| {
+                has_contracts.contains(&v)
+            }) {
+                Err("sender does not have these contracts.")?
+            }
+
+            let offer = OfferOf::<T> {
+                buyer,
+                sender,
+                contracts,
+                amount,
+                expired: T::Time::now() + expired,
+                state: OfferState::WAITING,
+            };
+
+            if let Some(current_offer) = <Offers<T>>::get(&offer_account) {
+                if current_offer.state == OfferState::WAITING {
+                    Err("this offer was already issued.")?
+                }
+            }
+
+            // lock amount
+            T::Currency::set_lock(
+                TRADING_ID, &offer.buyer, offer.amount,
+                <T as system::Trait>::BlockNumber::max_value(),
```

Comment:
What's about to lock it until expired block?

Reply:
We can not know exactly the block number that is after expired. So, calling `accept`/`reject`/`remove` methods unlocks it when expired.

Reply:
What’s about to set expiration in blocks?

Reply:
Yes.
Because the balances module only supported `Moment = BlockNumebr`.
https://github.com/paritytech/substrate/blob/master/frame/balances/src/lib.rs#L1312

Reply:
I mean when we known exportation block as parameter, it’s possible to set it here.

Reply:
Let’s estimate expired block on UI side?

Reply:
If we set expiration as block number,  we are uncertain and inconvenient. So, we should use the time.

Reply:
UI can estimate block number from user given time. On blockchain side working with block numbers much more safe when we talks about expiration. (Validators can manipulate timestamp as needed but can’t change count of blocks)

Reply:
I understand. I try to found how to estimate BlockNumber from Moment on-chain.

Reply:
Ah... not needs estimating on-chain.